network-mux/src/Network/Mux.hs

```diff
@@ -218,27 +216,27 @@ muxStart tracer (MuxApplication ptcls) bearer = do
                               (MiniProtocolException pnum pix pmode)
 
         jobAction run = do
-          chan    <- mkChannel
+          w       <- newTVarM BL.empty
+          let chan = muxChannel tracer tq (Wanton w)
+                                pnum pmode
+                                (selectQueue pmode)
           _result <- run chan
-          mpsJobExit cnt
+          mpsJobExit w
           return (MiniProtocolShutdown pnum pix pmode)
 
-        mkChannel = muxChannel tracer tq pnum pmode
-                               (selectQueue pmode) cnt
-
         selectQueue ModeInitiator = initQ
         selectQueue ModeResponder = respQ
 
-    -- cnt represent the number of SDUs that are queued but not yet sent.  Job
-    -- threads will be prevented from exiting until all SDUs have been
+    -- The Wanton w is the SDUs that are queued but not yet sent for this job.
+    -- Job threads will be prevented from exiting until all their SDUs have been
     -- transmitted unless an exception/error is encounter. In that case all
     -- jobs will be cancelled directly.
-    mpsJobExit :: StrictTVar m Int -> m ()
-    mpsJobExit cnt = do
+    mpsJobExit :: StrictTVar m BL.ByteString -> m ()
+    mpsJobExit w = do
```

Comment:
I agree with this fix but I'm worried that it is premature.
I thought that we would still bring down the mux bearer in case any mux protocol exited which in turn would cause the bearer to be torn down and all unsent messages for any other miniprotocol to not be sent out.

Reply:
I'll drop the last patch from this PR. We can include it into the next batch of changes.